Publishing via maven-publish plugin fails in Gradle 3.5 RC1

### Current Behavior

Publishing to a remote HTTP repository via the 'maven-publish' plugin results in the following error: 

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':scans-plugin:publishPluginPublicationToPublicRepository'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:98)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:68)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.execute(DefaultTaskGraphExecuter.java:236)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.execute(DefaultTaskGraphExecuter.java:228)
        at org.gradle.internal.Transformers$4.transform(Transformers.java:169)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:106)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:61)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:228)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:215)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:77)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:58)
        at org.gradle.execution.taskgraph.ParallelTaskPlanExecutor.process(ParallelTaskPlanExecutor.java:50)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:113)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
        at org.gradle.initialization.DefaultGradleLauncher$RunTasksAction.execute(DefaultGradleLauncher.java:230)
        at org.gradle.initialization.DefaultGradleLauncher$RunTasksAction.execute(DefaultGradleLauncher.java:227)
        at org.gradle.internal.Transformers$4.transform(Transformers.java:169)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:106)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:56)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:161)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:112)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:95)
        at org.gradle.launcher.exec.GradleBuildController.run(GradleBuildController.java:66)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:41)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:75)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:49)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:49)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:31)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:67)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
        at org.gradle.util.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:60)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:72)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.gradle.api.artifacts.PublishException: Failed to publish publication 'plugin' to repository 'public'
        at org.gradle.api.publish.internal.PublishOperation.run(PublishOperation.java:38)
        at org.gradle.api.publish.maven.tasks.PublishToMavenRepository.doPublish(PublishToMavenRepository.java:79)
        at org.gradle.api.publish.maven.tasks.PublishToMavenRepository.publish(PublishToMavenRepository.java:75)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:141)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:123)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:692)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:675)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.execute(ExecuteActionsTaskExecuter.java:115)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.execute(ExecuteActionsTaskExecuter.java:109)
        at org.gradle.internal.Transformers$4.transform(Transformers.java:169)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:106)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:56)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:90)
        ... 73 more
Caused by: org.gradle.api.GradleException: Failed to retrieve remote metadata com.gradle:build-scan-plugin/maven-metadata.xml: Could not transfer metadata com.gradle:build-scan-plugin/maven-metadata.xml from/to remote (http://docker:8081/artifactory/test-maven-repo): Could not get resource 'com/gradle/build-scan-plugin/maven-metadata.xml'
        at org.gradle.api.publication.maven.internal.action.AbstractMavenPublishAction.publish(AbstractMavenPublishAction.java:101)
        at org.gradle.api.publication.maven.internal.action.MavenDeployAction.publish(MavenDeployAction.java:35)
        at org.gradle.api.publish.maven.internal.publisher.MavenRemotePublisher$GradleWagonMavenDeployAction.publish(MavenRemotePublisher.java:87)
        at org.gradle.api.publish.maven.internal.publisher.AbstractMavenPublisher.execute(AbstractMavenPublisher.java:70)
        at org.gradle.api.publish.maven.internal.publisher.AbstractMavenPublisher.publish(AbstractMavenPublisher.java:47)
        at org.gradle.api.publish.maven.internal.publisher.StaticLockingMavenPublisher.publish(StaticLockingMavenPublisher.java:38)
        at org.gradle.api.publish.maven.internal.publisher.ValidatingMavenPublisher.publish(ValidatingMavenPublisher.java:48)
        at org.gradle.api.publish.maven.tasks.PublishToMavenRepository$1.publish(PublishToMavenRepository.java:85)
        at org.gradle.api.publish.internal.PublishOperation.run(PublishOperation.java:36)
        ... 92 more
Caused by: org.sonatype.aether.deployment.DeploymentException: Failed to retrieve remote metadata com.gradle:build-scan-plugin/maven-metadata.xml: Could not transfer metadata com.gradle:build-scan-plugin/maven-metadata.xml from/to remote (http://docker:8081/artifactory/test-maven-repo): Could not get resource 'com/gradle/build-scan-plugin/maven-metadata.xml'
        at org.sonatype.aether.impl.internal.DefaultDeployer.upload(DefaultDeployer.java:403)
        at org.sonatype.aether.impl.internal.DefaultDeployer.deploy(DefaultDeployer.java:292)
        at org.sonatype.aether.impl.internal.DefaultDeployer.deploy(DefaultDeployer.java:211)
        at org.sonatype.aether.impl.internal.DefaultRepositorySystem.deploy(DefaultRepositorySystem.java:443)
        at org.gradle.api.publication.maven.internal.action.MavenDeployAction.publishArtifacts(MavenDeployAction.java:69)
        at org.gradle.api.publication.maven.internal.action.AbstractMavenPublishAction.publish(AbstractMavenPublishAction.java:99)
        ... 100 more
Caused by: org.sonatype.aether.transfer.MetadataTransferException: Could not transfer metadata com.gradle:build-scan-plugin/maven-metadata.xml from/to remote (http://docker:8081/artifactory/test-maven-repo): Could not get resource 'com/gradle/build-scan-plugin/maven-metadata.xml'
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$3.wrap(WagonRepositoryConnector.java:930)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$3.wrap(WagonRepositoryConnector.java:920)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$GetTask.run(WagonRepositoryConnector.java:669)
        at org.sonatype.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:60)
        ... 3 more
Caused by: org.apache.maven.wagon.TransferFailedException: Could not get resource 'com/gradle/build-scan-plugin/maven-metadata.xml'
        at org.gradle.api.publication.maven.internal.wagon.RepositoryTransportDeployWagon.get(RepositoryTransportDeployWagon.java:84)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$GetTask.run(WagonRepositoryConnector.java:601)
        ... 4 more
Caused by: java.lang.IllegalStateException: No operation is currently running.
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.getCurrentOperation(DefaultBuildOperationExecutor.java:49)
        at org.gradle.internal.resource.transport.BuildOperationExternalResource.createBuildOperationDetails(BuildOperationExternalResource.java:124)
        at org.gradle.internal.resource.transport.BuildOperationExternalResource.writeTo(BuildOperationExternalResource.java:72)
        at org.gradle.api.publication.maven.internal.wagon.RepositoryTransportWagonAdapter.getRemoteFile(RepositoryTransportWagonAdapter.java:45)
        at org.gradle.api.publication.maven.internal.wagon.RepositoryTransportDeployWagon.get(RepositoryTransportDeployWagon.java:75)
        ... 5 more
```

### Context
Essentially, Maven publishing is broken in Gradle 3.5. This is a result of the new addition of build operations for remote artifact downloads. What triggers this is the maven publishing code fetching the existing `maven-metadata.xml` file. Because of Maven goofiness this is actually done in a separate thread spawned by the Maven wagon code. Because of this the lookup for the current build operation fails, due to this being stored in a `ThreadLocal`.

### Steps to Reproduce (for bugs)
Any build which publishes to a remote HTTP repository using the 'maven-publish' plugin will fail. Failure will only happen however when publishing an existing module, as the first publication will create the `maven-metadata.xml` (it won't have existed yet). I expect this is the reason why this got past our test coverage.

### Possible Fixes

We have a couple possible fixes here:

1. Change `DefaultBuildOperationExecutor` to use an `InheritableThreadLocal` instead.
   
    It's not completely clear to me whether this will introduce a concurrency issue though. From what I understand child threads simply inherit initial values from the parent. So concurrent child threads that create their own build operations (and therefor call `currentOperation.set()`) should not affect the parent thread?

2. Don't explicitly set the parent build operation in `BuildOperationExternalResource.createBuildOperationDetails()`. There is actually no need to do this anyhow as the following code handles setting the parent for us. 

   https://github.com/gradle/gradle/blob/7402b407a7176167e18857c4d92a5bca151ab028/subprojects/core/src/main/java/org/gradle/internal/progress/DefaultBuildOperationExecutor.java#L72-L72

The downside of option #2 is that the parent won't actually be set correctly on this build operation. So in the case of a UI like Buildship, this operation won't be correctly attributed to the publication task as should be the case. For build scans this has no effect as we don't show the operation hierarchy.

--- a/subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transport/BuildOperationExternalResource.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transport/BuildOperationExternalResource.java
@@ -121,7 +121,6 @@ public class BuildOperationExternalResource implements ExternalResource {
     private BuildOperationDetails createBuildOperationDetails() {
         ExternalResourceMetaData metaData = getMetaData();
         DownloadBuildOperationDescriptor downloadBuildOperationDescriptor = new DownloadBuildOperationDescriptor(metaData.getLocation(), metaData.getContentLength(), metaData.getContentType());
-        BuildOperationDetails buildOperationDetails = BuildOperationDetails.displayName("Download " + metaData.getLocation().toString()).parent(buildOperationExecutor.getCurrentOperation()).operationDescriptor(downloadBuildOperationDescriptor).build();
-        return buildOperationDetails;
+        return BuildOperationDetails.displayName("Download " + metaData.getLocation().toString()).operationDescriptor(downloadBuildOperationDescriptor).build();
     }
 }

--- a/subprojects/dependency-management/src/test/groovy/org/gradle/internal/resource/transport/BuildOperationExternalResourceTest.groovy
+++ b/subprojects/dependency-management/src/test/groovy/org/gradle/internal/resource/transport/BuildOperationExternalResourceTest.groovy
@@ -63,11 +63,9 @@ class BuildOperationExternalResourceTest extends Specification {
         def resource = new BuildOperationExternalResource(buildOperationExecuter, delegate)
 
         1 * delegate.getMetaData() >> metaData
-        1 * buildOperationExecuter.getCurrentOperation() >> operation
         1 * buildOperationExecuter.run(_, _) >> { details, action ->
             invokeAction(action, Mock(BuildOperationContext))
 
-            assert details.parent == operation
             assert details.name == "Download http://some/uri"
             assert details.displayName == "Download http://some/uri"
 
@@ -99,12 +97,10 @@ class BuildOperationExternalResourceTest extends Specification {
         def buildOperationExecuter = Mock(BuildOperationExecutor)
         def uri = new URI("http://some/uri")
         def metaData = new DefaultExternalResourceMetaData(uri, 0, 1024)
-        def operation = Mock(BuildOperationExecutor.Operation)
         def resource = new BuildOperationExternalResource(buildOperationExecuter, delegate)
         def buildOperationContext = Mock(BuildOperationContext)
 
         1 * delegate.getMetaData() >> metaData
-        1 * buildOperationExecuter.getCurrentOperation() >> operation
         1 * buildOperationExecuter.run(_, _) >> { details, action ->
             invokeAction(action, buildOperationContext)
         }

--- a/subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishHttpIntegTest.groovy
+++ b/subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishHttpIntegTest.groovy
@@ -216,6 +216,52 @@ class MavenPublishHttpIntegTest extends AbstractMavenPublishIntegTest {
         module.rootMetaData.versions == ["2"]
     }
 
+    @Issue("gradle/gradle#1641")
+    def "can publish a new version of a module already present in the target repository"() {
+        given:
+        buildFile << publicationBuild(version, group, mavenRemoteRepo.uri)
+        expectModulePublish(module)
+
+        when:
+        succeeds 'publish'
+
+        and:
+        buildFile.text = publicationBuild("3", group, mavenRemoteRepo.uri)
+        module = mavenRemoteRepo.module(group, name, "3")
+
+        then:
+        module.artifact.expectPut()
+        module.artifact.sha1.expectPut()
+        module.artifact.md5.expectPut()
+        module.pom.expectPut()
+        module.pom.sha1.expectPut()
+        module.pom.md5.expectPut()
+
+        and:
+        module.rootMetaData.expectGet()
+        module.rootMetaData.sha1.expectGet()
+        module.rootMetaData.expectGet()
+        module.rootMetaData.sha1.expectGet()
+        module.rootMetaData.expectPut()
+        module.rootMetaData.sha1.expectPut()
+        module.rootMetaData.md5.expectPut()
+
+        and:
+        succeeds 'publish'
+
+        then:
+        def localPom = file("build/publications/maven/pom-default.xml").assertIsFile()
+        def localArtifact3 = file("build/libs/publish-3.jar").assertIsFile()
+
+        module.pomFile.assertIsCopyOf(localPom)
+        module.pom.verifyChecksums()
+        module.artifactFile.assertIsCopyOf(localArtifact3)
+        module.artifact.verifyChecksums()
+
+        module.rootMetaData.verifyChecksums()
+        module.rootMetaData.versions == ["2", "3"]
+    }
+
     private String publicationBuild(String version, String group, URI uri, PasswordCredentials credentials = null) {
         String credentialsBlock = credentials ? """
                         credentials{

--- a/subprojects/maven/src/integTest/groovy/org/gradle/integtests/publish/maven/MavenPublishIntegrationTest.groovy
+++ b/subprojects/maven/src/integTest/groovy/org/gradle/integtests/publish/maven/MavenPublishIntegrationTest.groovy
@@ -727,4 +727,71 @@ uploadArchives {
         failure.assertHasCause "Could not publish configuration 'archives'"
         failure.assertThatCause(containsString('Cannot publish a directory'))
     }
+
+    @Issue("gradle/gradle#1641")
+    def "can publish a new version of a module already present in the target repository"() {
+        given:
+        server.start()
+        def mavenRemoteRepo = new MavenHttpRepository(server, "/repo", mavenRepo)
+        def group = 'org.gradle'
+        def name = 'publish'
+
+        and:
+        settingsFile << "rootProject.name = '$name'"
+        buildFile << """
+            apply plugin: 'java'
+            apply plugin: 'maven'
+            group = '$group'
+            uploadArchives {
+                repositories {
+                    mavenDeployer {
+                        repository(url: "${mavenRemoteRepo.uri}")
+                    }
+                }
+            }
+        """.stripIndent()
+
+        and:
+        def module1 = mavenRemoteRepo.module(group, name, '1')
+        module1.artifact.expectPut()
+        module1.artifact.sha1.expectPut()
+        module1.artifact.md5.expectPut()
+        module1.rootMetaData.expectGetMissing()
+        module1.rootMetaData.expectPut()
+        module1.rootMetaData.sha1.expectPut()
+        module1.rootMetaData.md5.expectPut()
+        module1.pom.expectPut()
+        module1.pom.sha1.expectPut()
+        module1.pom.md5.expectPut()
+
+        when:
+        succeeds 'uploadArchives', '-Pversion=1'
+
+        then:
+        module1.rootMetaData.verifyChecksums()
+        module1.rootMetaData.versions == ["1"]
+
+        and:
+        def module2 = mavenRemoteRepo.module(group, name, '2')
+        module2.artifact.expectPut()
+        module2.artifact.sha1.expectPut()
+        module2.artifact.md5.expectPut()
+        module2.pom.expectPut()
+        module2.pom.sha1.expectPut()
+        module2.pom.md5.expectPut()
+
+        and:
+        module2.rootMetaData.expectGet()
+        module2.rootMetaData.sha1.expectGet()
+        module2.rootMetaData.expectPut()
+        module2.rootMetaData.sha1.expectPut()
+        module2.rootMetaData.md5.expectPut()
+
+        when:
+        succeeds 'uploadArchives', '-Pversion=2'
+
+        then:
+        module2.rootMetaData.verifyChecksums()
+        module2.rootMetaData.versions == ["1", "2"]
+    }
 }

--- a/subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r35/BuildProgressCrossVersionSpec.groovy
+++ b/subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r35/BuildProgressCrossVersionSpec.groovy
@@ -26,6 +26,7 @@ import org.gradle.test.fixtures.server.http.MavenHttpRepository
 import org.gradle.test.fixtures.server.http.RepositoryHttpServer
 import org.gradle.tooling.ProjectConnection
 import org.junit.Rule
+import spock.lang.Issue
 
 @ToolingApiVersion(">=2.5")
 @TargetGradleVersion(">=3.5")
@@ -183,6 +184,70 @@ class BuildProgressCrossVersionSpec extends ToolingApiSpecification {
             LOGGER.warn("Unable to kill daemon(s)", ex);
         }
 
+    }
+
+    @Issue("gradle/gradle#1641")
+    def "generates download events during maven publish"() {
+        given:
+        toolingApi.requireIsolatedUserHome()
+        def module = mavenHttpRepo.module('group', 'publish', '1')
+
+        // module is published
+        module.publish()
+
+        // module will be published a second time via 'maven-publish'
+        module.artifact.expectPut()
+        module.artifact.sha1.expectPut()
+        module.artifact.md5.expectPut()
+        module.pom.expectPut()
+        module.pom.sha1.expectPut()
+        module.pom.md5.expectPut()
+        module.rootMetaData.expectGet()
+        module.rootMetaData.sha1.expectGet()
+        module.rootMetaData.expectGet()
+        module.rootMetaData.sha1.expectGet()
+        module.rootMetaData.expectPut()
+        module.rootMetaData.sha1.expectPut()
+        module.rootMetaData.md5.expectPut()
+
+        settingsFile << 'rootProject.name = "publish"'
+        buildFile << """
+            apply plugin: 'java'
+            apply plugin: 'maven-publish'
+            version = '1'
+            group = 'group'
+
+            publishing {
+                repositories {
+                    maven { url "${mavenHttpRepo.uri}" }
+                }
+                publications {
+                    maven(MavenPublication) {
+                        from components.java
+                    }
+                }
+            }
+        """
+
+        when:
+        def events = ProgressEvents.create()
+
+        withConnection {
+            ProjectConnection connection ->
+                connection.newBuild().forTasks('publish')
+                    .addProgressListener(events).run()
+        }
+
+        then:
+        def roots = events.operations.findAll { it.parent == null }.collect { it.descriptor.name }
+
+        roots == [
+            "Run build",
+            "Download ${module.rootMetaData.uri}",
+            "Download ${module.rootMetaData.sha1.uri}",
+            "Download ${module.rootMetaData.uri}",
+            "Download ${module.rootMetaData.sha1.uri}"
+        ]
     }
 
     def "generate events for task actions"() {

--- a/subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r35/BuildProgressCrossVersionSpec.groovy
+++ b/subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r35/BuildProgressCrossVersionSpec.groovy
@@ -187,6 +187,7 @@ class BuildProgressCrossVersionSpec extends ToolingApiSpecification {
     }
 
     @Issue("gradle/gradle#1641")
+    @LeaksFileHandles
     def "generates download events during maven publish"() {
         given:
         toolingApi.requireIsolatedUserHome()

--- a/subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ProgressEvents.groovy
+++ b/subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ProgressEvents.groovy
@@ -75,7 +75,8 @@ class ProgressEvents implements ProgressListener {
 
                     // Display name should be mostly unique
                     if (!skipValidation && uniqueBuildOperation(descriptor)) {
-                        if (descriptor.displayName in ['Configure settings', 'Configure build', 'Calculate task graph', 'Run tasks']) {
+                        if (descriptor.displayName in ['Configure settings', 'Configure build', 'Calculate task graph', 'Run tasks']
+                            || descriptor.displayName.contains('/maven-metadata.xml')) {
                             // Ignore this for now
                         } else {
                             def duplicateName = operations.find({ it.descriptor.displayName == descriptor.displayName })

